enhancements/cluster-logging/cluster-logging-elasticsearch-proxy.md

```diff
@@ -0,0 +1,121 @@
+
+creation-date: 2019-09-19
+last-updated: 2019-09-19
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# cluster-logging-elasticsearch-proxy
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The `elasticsearch-proxy` is a component that fronts Elasticsearch deployments for OpenShift Logging and is a replacement to the existing plugin `openshift-elasticsearch-plugin`. It provides authentication and authorization to control access to Elasticsearch.  Additionally, it handles logic specific to OpenShift logging to control access to stored data.  This is also the component that provides multi-tenancy access for Kibana, which allows regular users to see logs only from namespaces to which they belong, and allows admins to see logs from all namespaces and infrastructure/system logs.
+
+## Motivation
+
+The primary motivation for migrating logic to a component external to Elasticsearch deployments is to allow the development team to more quickly adopt newer releases of Elasticsearch.  This migration is proposed to coincide with forthcoming Elasticsearch v5 to v6 upgrade work.  The proxy will interface to the Elasticsearch cluster via its REST API which has shown to more stable then the plugin API.  The team has completed two major upgrades and the `openshift-elasticsearch-plugin` has shown to be specifically challenging to move to later versions.  Moving the multi-tenant functionality into the proxy and out of a plugin will additionally allow us to take advantage of several other community plugins without needing to modify them to our needs.
+
+### Goals
+
+The goals of this proposal are:
+
+* Provide a functional, "drop-in" replacement to the `openshift-elasticsearch-plugin` that supports Elasticsearch v6
+* Simplify the building and releasing of OpenShift Logging multi-tenant functionality
+
+We will be successful when our regression tests confirm the `elasticsearch-proxy` provides the same functionality as `openshift-elasticsearch-plugin`
+
+### Non-Goals
+
+## Proposal
+
+### Implementation Details
+
+The basic design of the proxy is to utilize a chain-or-responsibility to isolate specific aspects of the workflow:
+
+![https://docs.google.com/drawings/d/1dOm3TEBE8MTYmC_r1lBA7ZYVHZnEv_xg1bIBCs_wYhQ/edit](images/elasticsearch-proxy.png)
+
+* Authorization and authentication - Utilize the bearer token provided with the request to ensure the request originates with a recognized OpenShift token 
+* User info - Fetch information about a user (e.g. name, roles) to allow dynamic generation of permissions in ElasticSearch
+* Kibana Seeding - Populate a user's index to facilitate their ability to query logs
```

Comment:
is it not possible for us to do something more like prom-proxy, where we append a namespace to all the queries being performed, and thus don't need elastic to perform any access restrictions?  (so long as we then do not allow users to have direct access to elastic, but rather can only submit queries via our proxy).

Reply:
yes.  In order for us to simply the mods or proxy might do, however we need:
* An established log query API that is simple and  technology agnostic
* Require admins to grant privileges to users who need the ES api or who want to use Kibana to force them to take on the permissions burden

We tried to simplify this burden in the original release by doing some kibana request manipulation and seeding index patterns.  This has proven to be a primary source of the issues related to Kibana for which it isn't the best inteface for users who may just want to 'grep logs'

Reply:
tldr:  as long as we support multitenant kibana the proxy can't "just append namespaces" because of the complexity of the queries coming from kibana.  The proxy needs to work in conjunction with search guard or open distro to enforce access restrictions (with the proxy indicate what access is allowed, and search guard enforcing it on the query/results).